docs/explanation/charm-architecture.md

```diff
@@ -0,0 +1,18 @@
+# Charm architecture
+
+A [Juju](https://juju.is/) [charm](https://juju.is/docs/olm/charmed-operators) is a set of Python scripts and libraries that operates some software. This charm operates a set of [GitHub self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners) while handling the security and hardware management.
+
+Conceptually, the charm can be divided into the following:
+
+- Management of LXD ephemeral virtual machines to host [ephemeral self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/autoscaling-with-self-hosted-runners#using-ephemeral-runners-for-autoscaling).
+- Management of a [Python web service for checking GitHub repository settings](https://github.com/canonical/repo-policy-compliance).
+
+## LXD ephemeral virtual machines
+
+Using LXD ephemeral virtual machines to host the self-hosted runners ensures the environment is clean and isolated. The charm spawns a set of virtual machines with resources configured according to the charm configurations. The self-hosted runners are started with the ephemeral option and would clean up itself once it has taken a job from GitHub. At the end of the execution of the self-hosted runners application, the LXD ephemeral virtual machines would stop itself and clean up its resources. Freeing the resources to be used by future instances of virtual machines.
+
+## GitHub repository setting check
+
+A [flask application](https://flask.palletsprojects.com/) hosted on [gunicorn](https://gunicorn.org/) provides RESTful HTTP API to check the settings of GitHub repository. This is intended to ensure the GitHub repository settings does not allow for execution of untrusted code on the self-hosted runners managed by this charm.
```

Comment:
s/provides RESTful HTTP API to check the settings of GitHub repository. This is intended to ensure the GitHub repository settings does not allow for execution of untrusted code on the self-hosted runners managed by this charm./provides a RESTful HTTP API to check the settings of GitHub repository. This ensures the GitHub repository settings do not allow the execution of untrusted code on the self-hosted runners.

Reply:
I think we need to be very careful with wording around untrusted code. Having these repository settings doesn't prevent the execution of untrusted code, it only ensures changes are reviewed by project maintainers before being run. If a project maintainer let's untrusted code run, it will run :)

Reply:
Remove "untrusted code", replaced with "code not reviewed by maintainers"